---
title: "VDJ trimming/insertion regression by SNP--P value comparisons"
output:
  html_document:
    fig_crop: no
    toc: true
    toc_depth: 2
    dev: 'svg'
---

Here, we are plotting the top 100 p-values from Phil's regression analysis (no gene conditioning, using one trimming value per subject per trimming type) on the x-axis. On the y-axis are the p-values from my latest protocol where we are conditioning on the gene, by subject, and weighting the regression accordingly. In theory, this new protocol should get at the actual effect of the snp on trimming (without the potential influence of the snp effect on gene choice).

```{r echo=FALSE, message=FALSE, warning=FALSE}
library(knitr)
library(Cairo)
knitr::opts_chunk$set(dev="CairoPNG")
knitr::opts_chunk$set(fig.width=12, fig.height=8) 
knitr::opts_chunk$set(fig.path = "figures/")

source("manha_visualization.R")
source("find_correlated_snps.R")
source("make_snp_file.R")

pre_regression_snp_list = as.data.table(read.table("../_ignore/phil_sig_snps/top100_trimming_pvals_for_maggie.tsv", sep = "\t", fill=TRUE, header = TRUE))[-c(1:4)]

phil_snp_list = combine_trimming_regression_dfs(pre_regression_snp_list, condensing = 'by_gene', trimming_type_list = c('v_trim', 'd0_trim', 'd1_trim', 'j_trim', "vd_insert", "vj_insert", "dj_insert"))

artemis = snp_file(chromosome = 10, position1= 14397359, position2= 15454432)
tdt = snp_file(chromosome = 10, position1= 95804409, position2= 96838564)
mhc = snp_file(chromosome = 6, position1= 0, position2= 4050069)
rag = snp_file(chromosome = 11, position1= 36010709, position2= 37093156)
```


```{r echo=TRUE, warning=FALSE}
compare_phil_regression_list(phil_snp_list)
```

