---
title: "VDJ trimming/insertion regression by SNP visualizations"
output:
  html_document:
    fig_crop: no
    toc: true
    toc_depth: 2
    dev: 'svg'
---

```{r echo=FALSE, message=FALSE, warning=FALSE}
library(knitr)
library(Cairo)
knitr::opts_chunk$set(dev="CairoPNG")
knitr::opts_chunk$set(fig.width=12, fig.height=8) 
knitr::opts_chunk$set(fig.path = "figures/")

source("manha_visualization.R")
source("correlate_snps.R")
source("make_snp_file.R")

pre_regression_snp_list = as.data.table(read.table("../_ignore/phil_sig_snps/top100_trimming_pvals_for_maggie.tsv", sep = "\t", fill=TRUE, header = TRUE))[-c(1:4)]

phil_snp_list = combine_trimming_regression_dfs(pre_regression_snp_list, condensing = 'by_gene', trimming_type_list = c('v_trim', 'd0_trim', 'd1_trim', 'j_trim', "vd_insert", "vj_insert", "dj_insert"))

artemis = snp_file(chromosome = 10, position1= 14397359, position2= 15454432)
tdt = snp_file(chromosome = 10, position1= 95804409, position2= 96838564)
mhc = snp_file(chromosome = 6, position1= 0, position2= 4050069)
rag = snp_file(chromosome = 11, position1= 36010709, position2= 37093156)
```


```{r echo=TRUE, warning=FALSE}
productive = c("productive", "NOT_productive")
gene_region = c("artemis", "tdt", "rag", "mhc")

#correlated_snps_list = correlate_snps_ld(chromosome, artemis, cutoff = 0.85)

for (gene in gene_region){
    if (gene == 'mhc'){
        trimming_types = c("d0_trim", "d1_trim", "j_trim")
    } else {
        trimming_types = c("v_trim", "d0_trim", "d1_trim", "j_trim")
    }
    for (trim in trimming_types) {
        for (product in productive) {
            compile_data_manhattan(snp_meta_data = get(paste(gene)), snp_id_list = unique(get(gene)$snp), correlated_snps = 'False', productivity = product, varying_int = 'True', trim_type = trim, chromosome = unique(get(gene)$chr), correlate_snps = "False", gene_conditioning = 'True', weighting = 'True', gene_region = gene)
        }
    }
}
```