---
title: "VDJ trimming/insertion regression by SNPs across the genome"
output:
  html_document:
    fig_crop: no
    toc: true
    toc_depth: 2
    dev: 'svg'
fig_width: 26
fig_height: 10 
---

```{r echo=FALSE, message=FALSE, warning=FALSE}
library(knitr)
library(Cairo)
knitr::opts_chunk$set(dev="CairoPNG")
knitr::opts_chunk$set(fig.width=26, fig.height=10) 
knitr::opts_chunk$set(fig.path = "/home/mrussel2/tcr-gwas/trimming_regression/figures/genome/")

# Set bootstrap rerun count and pca structure correction variables globally
d_infer_trim = 'True'
d_infer_insert = 'False'
boot = 100
pca = 'True'
source("/home/mrussel2/tcr-gwas/trimming_regression/scripts/manha_visualization.R")
setDTthreads(1)
```
This file shows regressions across the entire genome. For these plots, the
second pass regressions are computing p values using:
```{r echo=FALSE, message=FALSE, warning=FALSE}
print(paste0('Bootstraps: ', boot))
print(paste0('PCA structure correction: ', pca))
```


All trimming regressions for artemis:
```{r ALL_trimming_plot_boot_pca_ARTEMIS, echo=TRUE, warning=FALSE}
manhattan_plot_cluster(trim_type = 'all_trim', random_effects = 'True', condensing = 'by_gene', d_infer = d_infer_trim, bootstrap_count = 0, plotting_cutoff = 4, gene_annotations = 'True', maf_cutoff = 0.05, bootstrap_rerun_count = boot, pca_structure_correction = pca, subset = 'artemis')
```


