---
title: "VDJ trimming/insertion regression by SNPs across the genome"
output:
  html_document:
    fig_crop: no
    toc: true
    toc_depth: 2
    dev: 'svg'
fig_width: 26
fig_height: 10 
---

This file shows regressions across the entire genome. 

```{r echo=FALSE, message=FALSE, warning=FALSE}
library(knitr)
library(Cairo)
knitr::opts_chunk$set(dev="CairoPNG")
knitr::opts_chunk$set(fig.width=26, fig.height=10) 
knitr::opts_chunk$set(fig.path = "/home/mrussel2/tcr-gwas/trimming_regression/figures/genome/")

# Set bootstrap rerun count and pca structure correction variables globally
bootstrap_rerun_count = 100
pca_structure_correction = 'True'
d_infer_trim = 'True'
d_infer_insert = 'False'

source("/home/mrussel2/tcr-gwas/trimming_regression/scripts/manha_visualization.R")
setDTthreads(1)
```

j trimming regressions:
```{r j_trimming_plot, echo=TRUE, warning=FALSE}
manhattan_plot_cluster(trim_type = 'j_trim', random_effects = 'True', condensing = 'by_gene', d_infer = d_infer_trim, bootstrap_count = 0, plotting_cutoff = 4, gene_annotations = 'True', maf_cutoff = 0.05, bootstrap_rerun_count, pca_structure_correction)
```

v trimming regressions:
```{r v_trimming_plot, echo=TRUE, warning=FALSE}
manhattan_plot_cluster(trim_type = 'v_trim', random_effects = 'True', condensing = 'by_gene', d_infer = d_infer_trim, bootstrap_count = 0, plotting_cutoff = 4, gene_annotations = 'True', maf_cutoff = 0.05, bootstrap_rerun_count, pca_structure_correction)
```

d0 trimming regressions:
```{r d0_trimming_plot, echo=TRUE, warning=FALSE}
manhattan_plot_cluster(trim_type = 'd0_trim', random_effects = 'True', condensing = 'by_gene', d_infer = d_infer_trim, bootstrap_count = 0, plotting_cutoff = 4, gene_annotations = 'True', maf_cutoff = 0.05, bootstrap_rerun_count,pca_structure_correction)
```

d1 trimming regressions:
```{r d1_trimming_plot, echo=TRUE, warning=FALSE}
manhattan_plot_cluster(trim_type = 'd1_trim', random_effects = 'True', condensing = 'by_gene', d_infer = d_infer_trim, bootstrap_count = 0, plotting_cutoff = 4, gene_annotations = 'True', maf_cutoff = 0.05, bootstrap_rerun_count, pca_structure_correction)
```

vd insertion regressions:
```{r vd_insert_plot, echo=TRUE, warning=FALSE}
manhattan_plot_cluster(trim_type = 'vd_insert', random_effects = 'False', condensing = 'by_patient', d_infer = d_infer_insert, bootstrap_count = 0, plotting_cutoff = 4, gene_annotations = 'True', maf_cutoff = 0.05, bootstrap_rerun_count, pca_structure_correction)
```

dj insertion regressions:
```{r dj_insert_plot, echo=TRUE, warning=FALSE}
manhattan_plot_cluster(trim_type = 'dj_insert', random_effects = 'False', condensing = 'by_patient', d_infer = d_infer_insert, bootstrap_count = 0, plotting_cutoff = 4, gene_annotations = 'True', maf_cutoff = 0.05, bootstrap_rerun_count, pca_structure_correction)
```

vj insertion regressions:
```{r vj_insert_plot, echo=TRUE, warning=FALSE}
manhattan_plot_cluster(trim_type = 'vj_insert', random_effects = 'False', condensing = 'by_patient', d_infer = d_infer_insert, bootstrap_count = 0, plotting_cutoff = 4, gene_annotations = 'True', maf_cutoff = 0.05, bootstrap_rerun_count, pca_structure_correction)
```

All trimming regressions:
```{r ALL_trimming_plot, echo=TRUE, warning=FALSE}
manhattan_plot_cluster(trim_type = 'all_trim', random_effects = 'True', condensing = 'by_gene', d_infer = d_infer_trim, bootstrap_count = 0, plotting_cutoff = 4, gene_annotations = 'True', maf_cutoff = 0.05, bootstrap_rerun_count, pca_structure_correction)
```

All insertion regressions:
```{r ALL_insertion_plot, echo=TRUE, warning=FALSE}
manhattan_plot_cluster(trim_type = 'all_insert', random_effects = 'False', condensing = 'by_patient', d_infer = d_infer_insert, bootstrap_count = 0, plotting_cutoff = 4, gene_annotations = 'True', maf_cutoff = 0.05, bootstrap_rerun_count, pca_structure_correction)
```

