---
title: "VDJ trimming/insertion regression by SNP visualizations"
output:
  html_document:
    fig_crop: no
    toc: true
    toc_depth: 2
    dev: 'svg'
---

```{r echo=FALSE, message=FALSE, warning=FALSE}
library(knitr)
library(Cairo)
knitr::opts_chunk$set(dev="CairoPNG")
knitr::opts_chunk$set(fig.width=12, fig.height=8) 
knitr::opts_chunk$set(fig.path = "/home/mrussel2/tcr-gwas/trimming_regression/figures/")

source("/home/mrussel2/tcr-gwas/trimming_regression/scripts/manha_visualization.R")
#source("/home/mrussel2/tcr-gwas/trimming_regression/scripts/find_correlated_snps.R")
source("/home/mrussel2/tcr-gwas/trimming_regression/scripts/make_snp_file.R")

artemis = snp_file(chromosome = 10, position1= 14397359, position2= 15454432)
tdt = snp_file(chromosome = 10, position1= 95804409, position2= 96838564)
mhc = snp_file(chromosome = 6, position1= 0, position2= 4050069)
rag = snp_file(chromosome = 11, position1= 36010709, position2= 37093156)
print('loaded data')
```


```{r echo=TRUE, warning=FALSE}
productive = c("productive", "NOT_productive")
gene_region = c("artemis", "tdt", "rag", "mhc")

#correlated_snps_list = correlate_snps_ld(chromosome, artemis, cutoff = 0.85)

for (gene in gene_region){
    trimming_types = c("v_trim", "d0_trim", "d1_trim", "j_trim")
    for (trim in trimming_types) {
        for (product in productive) {
            compile_data_manhattan(snp_meta_data = get(paste(gene)), snp_id_list = unique(get(gene)$snp), correlated_snps = 'False', paired_productive_snps = 'False', productivity = product, varying_int = 'True', trim_type = trim, chromosome = unique(get(gene)$chr), correlate_snps = "False", gene_conditioning = 'True', weighting = 'True', gene_region = gene)
            print('done')
        }
    }
}
```